File: 907-sum-of-subarray-minimums/907-sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        previousLess = []
        nextLess = []
        leftDistance = [-1] * len(arr)
        rightDistance = [-1] * len(arr)
        for i in range(len(arr)):
            while len(previousLess) != 0 and arr[previousLess[-1]] >= arr[i]:
                previousLess.pop()
            if len(previousLess) == 0:
                leftDistance[i] = i + 1
            else:
                leftDistance[i] = i - previousLess[-1]
            previousLess.append(i)
        
        for i in range(len(arr) - 1, -1, -1):
            while len(nextLess) != 0 and arr[nextLess[-1]] > arr[i]:
                nextLess.pop()
            if len(nextLess) == 0:
                rightDistance[i] = len(arr) - i
            else:
                rightDistance[i] = nextLess[-1] - i
            nextLess.append(i)
        ans = 0
        for i in range(len(arr)):
            ans += arr[i] * leftDistance[i] * rightDistance[i]
        return ans % (10**9 + 7)

File: 907-sum-of-subarray-minimums/907-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        HashMap<Integer, Integer> prevMin = new HashMap<>();
        HashMap<Integer, Integer> nextMin = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            while (!stack.isEmpty() && arr[stack.getLast()] >= arr[i]) {
                stack.removeLast();
            }
            if (stack.isEmpty()) {
                prevMin.put(i, -1);
            } else {
                prevMin.put(i, stack.getLast());
            }
            stack.addLast(i);
        }
        stack = new ArrayDeque<>();
        for (int i = arr.length - 1; i >= 0; i--) {
            while (!stack.isEmpty() && arr[stack.getLast()] > arr[i]) {
                stack.removeLast();
            }
            if (stack.isEmpty()) {
                nextMin.put(i, arr.length);
            } else {
                nextMin.put(i, stack.getLast());
            }
            stack.addLast(i);
        }
        
        long ans = 0;
        long mod = (int)Math.pow(10.0, 9.0) + 7;
        for (int i = 0; i < arr.length; i++) {
            long leftMin = i - prevMin.get(i);
            long rightMin = nextMin.get(i) - i;
            ans = (ans + arr[i] * leftMin * rightMin) % mod;
        }
        return (int)ans;
    }
}